pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,337 @@
+			members.Insert(groupAddress)
+			item = &mcastItem{
+				groupAddress: ip.IPv4StrToIPNet(groupAddress),
+				ruleIDs:      make(map[string]*uint16),
+			}
+			if groupAddresses.Has(groupAddress) {
+				// Just update
+				obj, exists, _ := c.mcastItemRuleCache.GetByKey(groupAddress)
+				if exists {
+					itemOrig := obj.(*mcastItem)
+					item.groupAddress = itemOrig.groupAddress
+					for ruleID, priority := range itemOrig.ruleIDs {
+						item.ruleIDs[ruleID] = priority
+					}
+				}
+			}
+			item.ruleIDs[igmpRule.ruleID] = igmpRule.priority
+			c.mcastItemRuleCache.Add(item)
+		}
+		for groupAddress := range groupAddresses {
+			if !members.Has(groupAddress) {
+				// Remove the stale ruleIDs
+				obj, exists, _ := c.mcastItemRuleCache.GetByKey(groupAddress)
+				if exists {
+					item = &mcastItem{
+						groupAddress: ip.IPv4StrToIPNet(groupAddress),
+						ruleIDs:      make(map[string]*uint16),
+					}
+					itemOrig := obj.(*mcastItem)
+					for ruleID, priority := range itemOrig.ruleIDs {
+						if ruleID != igmpRule.ruleID {
+							item.ruleIDs[ruleID] = priority
+						}
+					}
+					if len(item.ruleIDs) == 0 {
+						c.mcastItemRuleCache.Delete(itemOrig)
+					} else {
+						c.mcastItemRuleCache.Add(item)
+					}
+				}
+			}
+		}
+		c.ruleIDGroupAddressMap[igmpRule.ruleID] = members
+	case ruleDelete:
+		// Remove from cache
+		klog.V(2).InfoS("Delete rule", "igmpRule", igmpRule)
+		memberList := c.ruleIDGroupAddressMap[igmpRule.ruleID]
+		for member := range memberList {
+			obj, exists, _ := c.mcastItemRuleCache.GetByKey(member)
+			item = &mcastItem{
+				groupAddress: ip.IPv4StrToIPNet(member),
+				ruleIDs:      make(map[string]*uint16),
+			}
+			if exists {
+				itemOrig := obj.(*mcastItem)
+				for ruleID := range itemOrig.ruleIDs {
+					if ruleID != igmpRule.ruleID {
+						item.ruleIDs[ruleID] = itemOrig.ruleIDs[ruleID]
+					}
+				}
+				if len(item.ruleIDs) == 0 {
+					c.mcastItemRuleCache.Delete(itemOrig)
+				} else {
+					c.mcastItemRuleCache.Add(item)
+				}
+			}
+		}
+		delete(c.ruleIDGroupAddressMap, igmpRule.ruleID)
+	}
+	return nil
+}
+
+func newMulticastNetworkPolicyController(ofClient openflow.Client,
+	ifaceStore interfacestore.InterfaceStore,
+	ruleCache *ruleCache) (*multicastController, error) {
+	mcastItemRuleCache := cache.NewIndexer(groupAddressKey, cache.Indexers{
+		groupAddressIndex: groupAddressRuleIndexFunc,
+	})
+	mcastController := &multicastController{
+		ofClient:              ofClient,
+		ifaceStore:            ifaceStore,
+		ruleCache:             ruleCache,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "igmpRule"),
+		mcastItemRuleCache:    mcastItemRuleCache,
+		ruleIDGroupAddressMap: make(map[string]sets.String),
+	}
+	return mcastController, nil
+}
+
+func (c *multicastController) Validate(obj interface{}, groupAddress net.IP) (interface{}, error) {
```

Comment:
Where is `Validate` called? I cann't find any consumer.

Reply:
policy stats calls it for verifying whether IGMP report should be dropped in multicast packetIn

Reply:
I'm surprized networkpolicy behavior needs policy stats to enforce. It doesn't make sense from logically and techanically stats and policy should be enabled/disabled individually.
Why drop action can not be enforced by networkpolicy controller itself and keep a stats data in memory and let policy stats consume it?

Reply:
Sorry, it is not stat. IGMP report is handled by packetIn, so the block or allow operation is in packetIn.
And stat needs to calculate the IGMP packets dropped or allows by np. So for IGMP report, both np and stats are implemented in multicast packetIn

Reply:
Then where is the caller? I haven't found any from packetin handler

Reply:
it is in https://github.com/antrea-io/antrea/pull/3449, pkg/agent/multicast/mcast_discovery.go: addToPacketInStatsMap

Reply:
added the call of Validate in this pr